CEL: Enable regex pre-compilation, Add cost estimates for function extension libraries

jpbetz opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

- Enables CEL regex pre-compilation
- Adds cost calculations for all extension functions added to CEL by https://github.com/kubernetes/kubernetes/pull/108312, as well as the functions in the CEL "strings' extension library.

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/issues/107573

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

Yes, the only user visible feature of this change is the regex pre-compilation.

```release-note
CEL regex patterns in x-kubernetes-valiation rules are compiled when CRDs are created/updated if the pattern is provided as a string constant in the expression. Any regex compile errors are reported as a CRD create/update validation error.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language

jpbetz committed:
Add cost calculator for extensions functions, enable regex pre-compilation

jpbetz commented:
/priority important-soon
/triage accepted

jpbetz commented:
@TristonianJones would you be willing to do the first pass on this? The cost calculations got a bit fussy on this one.

jpbetz commented:
/assign @cici37
For review of runtime cost calculations.

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go:
```diff
@@ -141,7 +141,10 @@ func compileRule(rule apiextensions.ValidationRule, env *cel.Env) (compilationRe
-	prog, err := env.Program(ast, cel.EvalOptions(cel.OptOptimize))
+	prog, err := env.Program(ast,
+		cel.EvalOptions(cel.OptOptimize),
+		cel.OptimizeRegex(library.ExtensionLibRegexOptimizations...),
```
Do we want to add CostTracking(CostEstimator) for extension func cost cal later?

jpbetz replied:
Yes, there reason you're not seeing it here is because I want to integrate with https://github.com/kubernetes/kubernetes/pull/108419 after it merges. I should have made this clear in the description. I'll add an explanation now.

jpbetz replied:
This is all integrated now.

jpbetz commented:
/assign @liggitt 

This is stacked on https://github.com/kubernetes/kubernetes/pull/108419, but should be easy to review (just look at the HEAD commit). I'll rebase on master once https://github.com/kubernetes/kubernetes/pull/108419 merges.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost.go:
```diff
@@ -0,0 +1,250 @@
+	switch function {
+	case "isSorted", "sum", "max", "min", "indexOf", "lastIndexOf":
+		var cost uint64
+		if lister, ok := args[0].(traits.Lister); ok {
```
are we always guaranteed len(args) > 0? same question applies below

jpbetz replied:
The compiler does pre-check against the declared signatures, but I've added guards to be extra safe, and to be consistent across the estimated cost and runtime cost code.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost.go:
```diff
@@ -0,0 +1,250 @@
+		var cost uint64
+		if lister, ok := args[0].(traits.Lister); ok {
+			// Start with a base 1 cost for comparing each element in the list.
+			size := uint64(lister.Size().(types.Int))
```
are all of these casts guaranteed safe?

jpbetz replied:
Yes, the size ones are safe. xref from cel-go: https://github.com/google/cel-go/blob/9e07c9972efa837cb00dda458286bdf96eb15309/common/types/map.go#L650

cc @TristonianJones (incase there is any reason I should guard the check that is super subtle here).

TristonianJones replied:
You're not missing anything subtle here. The type of `Size` should be a `types.Int`. I'm trying to recall why I structured the return value that way and I'm having some trouble recalling because it was quite a while ago.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost.go:
```diff
@@ -0,0 +1,250 @@
+				case types.Bytes:
+					bytesCompared += uint64(len([]byte(vt)))
+				default:
+					break // not a string or bytes, so we don't need to traverse the whole list
```
would a literal list containing a variable hit this case? does this mean we'll only estimate cost for literal lists containing all literal strings?

jpbetz replied:
I've generalized this more and moved it into its own function.

liggitt reviewed (commented):
it would be good if someone with more context on the cel cost calculation could review as well (didn't see more than the one comment from cici, so I wasn't sure if she/tristan had reviewed this in detail)

my main questions are:
1. is there unit test coverage of the added calculations (scanned with `go test $whatever -coverprofile=coverage.out && go tool cover -html=coverage.out`)?
2. are the assumptions about args and type casts guaranteed safe?
3. if our calculations in these cost estimates needs to change in the future, could that make previously persisted CRDs be considered invalid? if so, this code needs to be flagged as impacting API compatibility, and/or there should be a plan for how to allow a no-op update to a CRD to succeed even if our cost estimates now think the cel expressions in it could be too expensive

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost_test.go:
```diff
@@ -0,0 +1,363 @@
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	_, details, err := prog.Eval(map[string]interface{}{})
```
Since we now switched to use ContextEval, would you mind to update here as well

jpbetz replied:
It's not needed here.

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost_test.go:
```diff
@@ -0,0 +1,363 @@
+}
+
+func testCost(t *testing.T, expr string, expectEsimatedCost checker.CostEstimate, expectRuntimeCost uint64) {
+	est := &CostEstimator{SizeEstimator: &testCostEstimator{}}
```
Maybe pass in library.CostEstimator to test?

jpbetz replied:
`CostEstimator` on this line is `library.CostEstimator`. The passed in estimator is just for size estimation mocking.

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost_test.go:
```diff
@@ -0,0 +1,363 @@
+	if estCost.Min != expectEsimatedCost.Min || estCost.Max != expectEsimatedCost.Max {
+		t.Errorf("Expected estimated cost of %d..%d but got %d..%d", expectEsimatedCost.Min, expectEsimatedCost.Max, estCost.Min, estCost.Max)
+	}
+	prog, err := env.Program(compiled, cel.CostTracking(est))
```
I feel there is a gap in test. Would it make more sense to keep the same configuration as in compilation.go for testing?(like to use the libraries.estimator?)

jpbetz replied:
I want to test this in isolation as much as possible to make it easy to reason about. Which part of the configuration from compilation.go would help here?

cici37 replied:
My concern is that the current test for library is pretty much isolated from the code path which might increase maintenance effort. Since the library is more general purpose as you said and we might port some of that into CEL in the future, lgtm.

jpbetz commented:
> is there unit test coverage of the added calculations (scanned with go test $whatever -coverprofile=coverage.out && go tool cover -html=coverage.out)?

Yes, the coverage from library/test_cost.go looks good to me here (all of CallCost and EstimateCallCost are covered).

> are the assumptions about args and type casts guaranteed safe?

I've added a few more guards to be extra safe. The compiler pre-checks the arg counts. The unchecked cases of `Size()`are guaranteed (cel-go makes the same assumption).

> if our calculations in these cost estimates needs to change in the future, could that make previously persisted CRDs be considered invalid? if so, this code needs to be flagged as impacting API compatibility, and/or there should be a plan for how to allow a no-op update to a CRD to succeed even if our cost estimates now think the cel expressions in it could be too expensive

I've added a warning to flag that the estimated costs are API impacting and should not be changed.  I've added a task to write up a plan for CEL cost ratcheting. I think are two parts to it: (1) if no CEL rules are changed in an update, they are always allowed, regardless of what the current cost calculations are and (2) if some of the CEL rules are updated, the per-expression limits for those rules apply and the total cost applies. I'd be comfortable committing to implementing this before promoting to beta.

TristonianJones commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost_test.go:
```diff
@@ -0,0 +1,364 @@
+	cases := []struct {
+		name               string
+		expr               string
+		expectEsimatedCost checker.CostEstimate
```
nit: s/Esimated/Estimated/

TristonianJones commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go:
```diff
@@ -1512,6 +1512,80 @@ func TestCostEstimation(t *testing.T) {
+				objType = withRule(objType, "self.str.split(self.separator) == []")
+				return &objType
+			},
+			expectedCalcCost: 629160,
```
nit: Try to be consistent between `expect` and `expected` prefixes on test structs.

TristonianJones commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go:
```diff
@@ -1512,6 +1512,80 @@ func TestCostEstimation(t *testing.T) {
+		},
 	}
 	for _, testCase := range cases {
 		t.Run(testCase.name, func(t *testing.T) {
```
Within the test case, I noticed that `expectedCalcCost` is only used with a non-zero `expectCalcCostExceedsLimit`. It's hard to reason about the intention of the value when it's present for cases with a limit of zero.

jpbetz replied:
That's true, the case I added doesn't set `expectCalcCostExceedsLimit` so it's uses the test flow where there is no limit. I'm hesitant to restructure the test suite as part of this PR, but I'll adjust a few things to try to clarify the intent.

cc @DangerOnTheRanger

TristonianJones replied:
Thanks. Appreciate it.

TristonianJones commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost.go:
```diff
@@ -0,0 +1,272 @@
+		if len(list.Path()) > 0 {
+			// Provide path if we have it so that a OpenAPIv3 maxLength validation can be looked up, if it exists
+			// for this node.
+			return &itemsNode{path: append(list.Path(), "@items"), t: lt.ElemType, expr: nil}
```
Be careful with the append since it can mutate in place, and I'm not sure that's desired here. Probably best to return a copy of the path from `list.Path()`

jpbetz replied:
Thank you, fixed.

TristonianJones commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost.go:
```diff
@@ -0,0 +1,272 @@
+			return &itemsNode{path: append(list.Path(), "@items"), t: lt.ElemType, expr: nil}
+		} else {
+			// Provide just the type if no path is available so that worst case size can be looked up based on type.
+			return &itemsNode{t: lt.ElemType, expr: nil}
```
Prefer using the accessors when working with the proto form: `lt.GetElemType()` here and elsewhere since the new golang proto api v2 uses them almost exclusively (or plans to).

TristonianJones commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost.go:
```diff
@@ -0,0 +1,272 @@
+			i := it.Next()
+			switch i.(type) {
+			case types.String, types.Bytes, traits.Lister, traits.Mapper:
+				cost += traversalCost(i)
```
I think you can continue summing up the `traversalCost(i)` here and don't need the switch and early return. In fact you probably don't need to seed the cost with the `size()` value since you would normally compute it via traversing the elements. Thoughts? Either way, it would be good to be consistent with the traversal cost below for maps.

TristonianJones commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost.go:
```diff
@@ -0,0 +1,272 @@
+		for it := vt.Iterator(); it.HasNext() == types.True; {
+			k := it.Next()
+			v := vt.Get(k)
+			cost += uint64(len(string(k.(types.String)))) + traversalCost(v)
```
Consider just summing up the keys and values with traversal cost.  Unclear if you need to seed the calculation with the map's size. Same idea for lists.

TristonianJones commented:
@jpbetz LGTM

cici37 commented:
/lgtm

liggitt commented:
/approve
